game-core/src/main/java/games/strategy/engine/framework/startup/launcher/ServerLauncher.java

```diff
@@ -300,11 +267,7 @@ private void saveAndEndGame(final INode node) {
     stopGame();
 
     final String message = "Connection lost to:" + node.getName() + " game is over.  Game saved to:" + f.getName();
-    if (headless) {
-      log.info(message);
-    } else {
-      SwingUtilities.invokeLater(() -> JOptionPane.showMessageDialog(JOptionPane.getFrameForComponent(ui), message));
-    }
+    launchAction.onEnd(message);
```

Comment:
Seeing an abstraction problem here, SRP violation perhaps. In one case we use `launchAction` to tell us the save file name, then we call it later to be a handler for an event. Very different actions.

IN another comment I suggest breaking up the object to two different ones, that perhaps could help here.